### PR TITLE
Add fast beam search decoding

### PR DESCRIPTION
**Pull request by pkufool**
This PR uses the GPU parallel decoding method in https://github.com/k2-fsa/k2/pull/926.

The decoding results and time used are as follows:

I used the same model as https://github.com/k2-fsa/icefall/pull/248. The decoding process can run with the max value of **max-duration=2000**, which means **several hundred streams** for not very long utterances.  You can see in the table, the decoding time seems not affect very much by the `beam`, `max-contexts`,  `max-states` (I think it is because we use a small decoding graph, i.e. k2.trivial_graph). So I suggest to use beam=4, max-contexts=4, max-states=8 here.

As for the WERs, it is not as good as the results in https://github.com/k2-fsa/icefall/pull/248 that decoded with `modified_beam_search`(which is 2.56||6.27).  I think one of the reasons is we use much larger batch-size, when I used `max-duration=1` to decode, I can got  WERs of 2.58 || 6.28.

Will try HLG graph to see if it can reduce the WERs any further.

<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

beam | max-contexts | max-states | WERs for test-clean | WERs for test-other | time for test-clean(seconds) | time for test-other(seconds)
-- | -- | -- | -- | -- | -- | --
1 | -1 | -1 | 2.62 | 6.34 | 50 | 47
3 | -1 | -1 | 2.61 | 6.29 |   |  
5 | -1 | -1 | 2.61 | 6.29 | 50 | 49
7 | -1 | -1 | 2.61 | 6.29 |   |  
9 | -1 | -1 | 2.61 | 6.29 |   |  
12 | -1 | -1 | 2.61 | 6.29 | 50 | 48
15 | -1 | -1 | 2.61 | 6.29 |   |  
15 | 1 | 1 | 2.62 | 6.37 | 50 | 49
15 | 2 | 2 | 2.62 | 6.31 |   |  
15 | 2 | 4 | 2.62 | 6.31 |   |  
15 | 2 | 6 | 2.62 | 6.31 |   |  
15 | 3 | 3 | 2.61 | 6.29 |   |  
15 | 3 | 6 | 2.61 | 6.29 |   |  
15 | 3 | 9 | 2.61 | 6.29 |   |  
15 | 4 | 4 | 2.61 | 6.29 |   |  
15 | 4 | 8 | 2.61 | 6.29 |   |  
15 | 4 | 12 | 2.61 | 6.29 |   |  
15 | 5 | 5 | 2.61 | 6.29 |   |  
15 | 5 | 10 | 2.61 | 6.29 |   |  
15 | 5 | 15 | 2.61 | 6.29 |   |  
15 | 6 | 6 | 2.61 | 6.29 |   |  
15 | 6 | 12 | 2.61 | 6.29 |   |  
15 | 6 | 18 | 2.61 | 6.29 |   |  
15 | 8 | 8 | 2.61 | 6.29 |   |  
15 | 8 | 16 | 2.61 | 6.29 |   |  
15 | 8 | 24 | 2.61 | 6.29 |   |  
15 | 10 | 10 | 2.61 | 6.29 |   |  
15 | 10 | 20 | 2.61 | 6.29 |   |  
15 | 10 | 30 | 2.61 | 6.29 | 52 | 48

</div>

### CODE REVIEW

**Comment by danpovey** [comment]
Great progress!!
I don't understand why the batch size would make a difference?  Is there anything in the algorithm that would behave differently when things are batched?
I would not necessarily regard WER differences of 0.02 or 0.03 as better or worse.  These  are well within the range of experimental noise.

**Comment by csukuangfj** [comment]
> I don't understand why the batch size would make a difference?

Would the padding in the conv1d from the conformer make a difference here?

File: egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py [hunk 1]
```diff
@@ -17,9 +17,86 @@
...
+    encoder_out: torch.Tensor,
+    encoder_out_lens: torch.Tensor,
+    beam: float,
+    max_states: int,
+    max_contexts: int,
+) -> List[int]:
+    """It limits the maximum number of symbols per frame to 1.
+
+    Args:
+      model:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py:40]: The order of the arg doc should match the order of definition.

File: egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py [hunk 2]
```diff
@@ -17,9 +17,86 @@
...
+
+def fast_beam_search(
+    decoding_graph: k2.Fsa,
+    model: Transducer,
+    encoder_out: torch.Tensor,
+    encoder_out_lens: torch.Tensor,
+    beam: float,
+    max_states: int,
+    max_contexts: int,
+) -> List[int]:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py:36]:
```suggestion
) -> List[List[int]]:
```

File: egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py [hunk 3]
```diff
@@ -17,9 +17,86 @@
...
+    """It limits the maximum number of symbols per frame to 1.
+
+    Args:
+      model:
+        An instance of `Transducer`.
+      encoder_out:
+        A tensor of shape (N, T, C) from the encoder. Support only N==1 for now.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py:43]:
```suggestion
        A tensor of shape (N, T, C) from the encoder.
```

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py [hunk 4]
```diff
@@ -17,9 +17,86 @@
...
+    B, T, C = encoder_out.shape
+
+    config = k2.RnntDecodingConfig(
+        vocab_size=vocab_size,
+        decoder_history_len=context_size,
+        beam=beam,
+        max_contexts=max_contexts,
+        max_states=max_states,
+    )
+    indivisual_streams = []
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless/beam_search.py:69]: indivisual -> individual

**Pull request merged**